src/mame/olivetti/att6300p.cpp

```diff
@@ -0,0 +1,1050 @@
+	// Bit 6 = IOCHK
+	// Bit 7 = MBMERR (MRD parity check)
+
+	if (BIT(m_dsw->read(), 12))
+		m_ctrlport_b |= 0x02;
+	else
+		m_ctrlport_b &= 0xfd;
+
+	return m_ctrlport_b;
+}
+
+void att6300p_state::chck_w(int state)
+{
+	m_chck_active = (state == 0);
+	if (m_chck_active)
+	{
+		if (!BIT(m_ctrlport_b, 6))
+		{
+			m_ctrlport_b |= 0x40;
+			update_nmi();
+		}
+	}
+	else if (BIT(m_ctrlport_a, 4))
+		m_ctrlport_b &= 0xbf;
+}
+
+void att6300p_state::nmi_enable_w(uint8_t data)
+{
+	m_nmi_enable = BIT(data, 7);
+	update_nmi();
+}
+
+void att6300p_state::update_nmi()
+{
+	m_nmi_active = (m_trapio_active ||
+	  (m_nmi_enable && BIT(m_ctrlport_b, 6) && !BIT(m_ctrlport_a, 4)));
+
+	if (m_nmi_active)
+		m_maincpu->set_input_line(INPUT_LINE_NMI, ASSERT_LINE);
+	else
+		m_maincpu->set_input_line(INPUT_LINE_NMI, CLEAR_LINE);
+}
+
+uint8_t att6300p_state::kbc_p1_r()
+{
+	return (~m_p2_out & (m_kbdata ? 3 : 1)) | BIT(m_ctrlport_a, 6) << 2;
+}
+
+void att6300p_state::kbc_p2_w(uint8_t data)
+{
+	m_p2_out = data;
+
+	m_keyboard->clock_w(!BIT(data, 0));
+	m_keyboard->data_w(!BIT(data, 1));
+
+	// Needed to ensure the keyboard MCU gets scheduled to respond to
+	// the clock change and output the data in time.
+	machine().scheduler().perfect_quantum(attotime::from_usec(50));
+
+	if (m_kb_int != BIT(data, 4))
+	{
+		m_kb_int = BIT(data, 4);
+		update_ir1();
+	}
+}
+
+void att6300p_state::kbcin_w(int state)
+{
+	m_kbdata = state;
+}
+
+void att6300p_state::zc01_cb(int state)
+{
+	m_ctc->trg2(state);
+
+	if (state)
+	{
+		// Timeslice interrupt
+		m_ts_int = true;
+		update_ir1();
+	}
+}
+
+void att6300p_state::zc02_cb(int state)
+{
+	if (state)
+	{
+		// Sanity timer expired.
+		m_sanity_active = true;
+		soft_reset();
+	}
+}
+
+void att6300p_state::soft_reset()
+{
+	set_protected_mode(false);
+	m_maincpu->reset();
+}
+
+void att6300p_state::configure_mapping(int sel)
+{
+	static const uint8_t reverse_bits[32] = {
+		0x00, 0x10, 0x08, 0x18, 0x04, 0x14, 0x0c, 0x1c,
+		0x02, 0x12, 0x0a, 0x1a, 0x06, 0x16, 0x0e, 0x1e,
+		0x01, 0x11, 0x09, 0x19, 0x05, 0x15, 0x0d, 0x1d,
+		0x03, 0x13, 0x0b, 0x1b, 0x07, 0x17, 0x0f, 0x1f,
+	};
+
+	bool rom_upper = (sel & 0x8) == 0 || m_protected;
+	int rom_region = (sel & 7) + (rom_upper ? 8 : 0);
+
+	uint8_t *rom = m_map_rom->base() + rom_region*32;
+
+	uint32_t table[32];
+	for (int i = 0; i < 32; i++)
+	{
+		// AA19:AA15 inputs to the PROM are reversed
+		int idx = reverse_bits[i];
+
+		// Redirect to a 32K region below 1M
+		table[i] = (rom[idx] >> 3)*32*1024;
+
+		if (rom[idx] & 0x04)
+		{
+			// Redirect to top area of RAM
+			table[i] |= 0xf00000;
+		}
+	}
+
+	m_mmu->set_mem_mapping(table);
+
+	m_mapping_sel = sel;
+}
+
+void att6300p_state::set_protected_mode(bool enabled)
+{
+	if (m_protected != enabled)
+	{
+		m_protected = enabled;
+		configure_mapping(m_mapping_sel);
+		m_mmu->set_protected_mode_enabled(enabled);
+	}
+}
+
+uint8_t att6300p_state::virt_r(offs_t offset)
+{
+	uint8_t reg = offset & 0xe0;
+
+	switch (reg)
```

Comment:
It's usually preferred to let the `address_map` handle the decoding if possible and use individual functions instead.